SCClassLibrary/External/Ableton/LinkClock.sc

```diff
@@ -11,6 +13,169 @@ LinkClock : TempoClock {
+						queriedAtBeat: msg[2],
+						syncMeter: msg[3].asBoolean,
+						beats: msg[4],
+						beatsPerBar: msg[5],
+						baseBarBeat: msg[6],
+						beatInBar: msg[7]
+					));
+				};
+			}, '/LinkClock/meterReply');
+			{
+				NetAddr.broadcastFlag = true;
+				// maybe some peers had to take a different port
+				(57120 .. 57127).do { |port|
+					NetAddr("255.255.255.255", port).sendMsg('/LinkClock/queryMeter', this.beats);
+				};
+				(timeout * this.tempo).wait;  // seconds, not beats
+				resp.free;
+				action.value(replies.values);
+			}.fork(this);
+		} {
+			// no peers, don't bother asking
+			// sched is needed to make sure Conditions waiting for this method still work
+			this.sched(0, { action.value(List.new) });
+		};
+	}
+
+	adjustMeterBase { |localBeats, remoteBeats, round = 1|
+		baseBarBeat = baseBarBeat + ((localBeats - remoteBeats) % beatsPerBar).round(round);
+	}
+
+	// if there are peers with syncMeter = true, align my clock to their common barline
+	// ignore peers with syncMeter = false
+	resyncMeter { |round, verbose = true|
+		var replies, cond = Condition.new, bpbs, baseBeats, denom;
+		var newBeatsPerBar, newBase;
+		fork {
+			this.queryMeter { |val|
+				replies = val;
+				cond.unhang;
+			};
+			cond.hang;
+			replies = replies.select { |reply| reply[\syncMeter] };
+			if(replies.size > 0) {
+				// verify beatsPerBar, and myBeats - beatInBar, are common across peers
+				// (if the network is glitchy, we may have replies from different trials.
+				// but the difference between query time and remote beatInBar should be
+				// consistent.)
+				// also calculate baseBarBeat common denominator
+				bpbs = Set.new;
+				baseBeats = Set.new;
+				denom = 1;
+				replies.do { |reply, i|
+					denom = denom lcm: reply[\baseBarBeat].asFraction[1];
+					bpbs.add(reply[\beatsPerBar]);
+					baseBeats.add(reply[\queriedAtBeat] - reply[\beatInBar]);
+				};
+				if(round.isNil) { round = denom.reciprocal };
+				// Now it gets tricky. We need to reduce baseBeats based on beatsPerBar.
+				// All 'syncMeter' peers should have the same beatsPerBar (size == 1).
+				// But, if something failed, maybe there are more.
+				// So we have to do a two-step check.
+				if(bpbs.size == 1) {
+					// 'choose' = easy way to get one item from an unordered collection
+					newBeatsPerBar = bpbs.choose;
+					// Sets, by definition, cannot hold multiple items of equivalent value
+					// so this automatically collapses to minimum size!
+					baseBeats = baseBeats.collect { |x| x.round(round) % newBeatsPerBar };
+					if(baseBeats.size == 1) {
+						// do not change beats! do not ever change beats here!
+						// calculate baseBarBeat such that my local beatInBar will match theirs
+						// myBeats = replies.choose[\queriedAtBeat].round(round);
+						// theirPhase = bibs.choose;  // should be only one
+						newBase = baseBeats.choose;
+						if(verbose) { "syncing meter to %, base = %\n".postf(bpbs.choose, newBase) };
```

Comment:
this could be `newBeatsPerBar`

Reply:
i'm generally skeptical of printing informative text by default. can you talk a little bit about this one?

Reply:
> this could be newBeatsPerBar

Good point -- and now that it's `.pop` (which modifies the set), it *has* to be.

> i'm generally skeptical of printing informative text by default.

There is one concrete reason to post. Suppose you already have one SC LinkClock with a meter sync object with beatsPerBar = 3, and on another sclang instance, you start another LinkClock, with meter sync, and default settings (beatsPerBar = 4). Meter sync demands that the new clock join the 3/4 meter (and, if the network is functioning, it will), but that isn't what the code explicitly requested. So the user may have thought she was asking for normal 4/4 time, but ended up with 3/4 (because meter sync means "your requested meter, or others' meter if available"). *Maybe* it's better to do that silently, but I think it's better to advise the user.

I could change this to `if(verbose and: { newBeatsPerBar != clock.beatsPerBar })` perhaps.